src/inspector/data.rs

```diff
@@ -0,0 +1,196 @@
+#[derive(Clone, Debug)]
+pub enum DataType {
+    Internal {
+        children: Vec<CapturedData>,
+        expanded: RwSignal<bool>,
+    },
+    Leaf,
+}
+#[derive(Clone, Debug)]
+pub struct CapturedData {
+    pub id: ViewId,
+    pub view_conf: Rc<CapturedView>,
+    pub ty: DataType,
+}
+
+impl CapturedData {
+    pub fn init_from_view(view: Rc<CapturedView>) -> Self {
+        if view.children.is_empty() {
+            Self {
+                id: view.id,
+                view_conf: view,
+                ty: DataType::Leaf,
+            }
+        } else {
+            let mut children = Vec::with_capacity(view.children.len());
+            for child in &view.children {
+                children.push(CapturedData::init_from_view(child.clone()));
+            }
+            let expanded = create_rw_signal(false);
+            Self {
+                id: view.id,
+                view_conf: view,
+                ty: DataType::Internal { children, expanded },
+            }
+        }
+    }
+    pub fn count_line(&self, id: ViewId, line: &mut usize) -> bool {
+        line.add_assign(1);
+        if self.id == id {
+            return true;
+        }
+        match &self.ty {
+            DataType::Internal { children, expanded } => {
+                if expanded.get_untracked() {
+                    for child in children {
+                        if child.count_line(id, line) {
+                            return true;
+                        }
+                    }
+                }
+            }
+            DataType::Leaf => {}
+        }
+        false
+    }
+
+    pub fn focus(&mut self, id: ViewId) -> bool {
+        if self.id == id {
+            return true;
+        }
+        match &mut self.ty {
+            DataType::Internal { children, expanded } => {
+                for child in children.iter_mut() {
+                    if child.focus(id) {
+                        expanded.set(true);
+                        return true;
+                    }
+                }
+            }
+            DataType::Leaf => {}
+        }
+        false
+    }
+
+    pub fn expanded(&self) -> Option<RwSignal<bool>> {
+        match &self.ty {
+            DataType::Internal { expanded, .. } => Some(*expanded),
+            DataType::Leaf => None,
+        }
+    }
+
+    /// contain self?
+    fn total(&self) -> usize {
+        match &self.ty {
+            DataType::Internal { expanded, children } => {
+                if expanded.get() {
+                    let mut total = 1;
+                    for child in children {
+                        total += child.total();
+                    }
+                    total
+                } else {
+                    1
+                }
+            }
+            DataType::Leaf => 1,
+        }
+    }
+
+    fn get_children(
+        &self,
+        next: &mut usize,
+        min: usize,
+        max: usize,
+        level: usize,
+    ) -> Vec<(usize, usize, CapturedData)> {
```

Comment:
It looks like both of these functions are relatively expensive to compute. Is there a reason that these aren't just computed and stored in the init function? Then the get children function could just return a slice without needing to build a new vec of children elements.